### PR TITLE
refactor: sqlalchemy

### PR DESCRIPTION
**Pull request by abulte**
This migrates the ORM from `dataset` to `sqlalchemy`. 

Main advantages are:
- migrations support (#8)
- better typing handling
- schema is now constrained, and thus more manageable
- foreign keys support in ORM

The strategy is to create an "initial migration" that provides iso-features after being applied, without the need for data migrations. I tried to create a SQLAlchemy schema that would generate a quasi-empty initial migration, but it became a game of whack-a-mole I couldn't win.

The initial migration should be applied once everywhere and then removed, it's meant to be ephemeral and does not handle the bootstrapping of an empty database (see comment in migration file).

Apart from unit tests, the migration has been validated in the following scenarios:
- load script works on an unmigrated local database (not needed in practice)
- migration works on a `demo` database restored from our deployment, and subsequent load script works
- `python cli.py init-db` works to bootstrap an empty database, and subsequent load script works

I've compared the output of metrics computation in every scenario (this should validate quite a few things in the process chain).

On local, you should remove the `dataset` dependency, reinstall `requirements.txt` and either:
- start from scratch (probably the best option, and made "easier" by the fact that the database name has changed)
- or run `ALEMBIC_ENV=(demo|prod) alembic upgrade head` to migrate an existing database

⚠️ The local databases are now `dashboard_backend` for demo and `dashboard_backend_prod` for prod. This allows easy restore from our deployment. Delete your docker container to have the new databases created (you will lose your existing databases) or migrate them yourself (the Postgres version has changed too, so this will be difficult). Also remember to change your local `DATABASE_URL` and `DATABASE_URL_PROD` env vars values.

### CODE REVIEW

**Review by at-github — APPROVED** [review]

File: models.py [hunk 1]
```diff
@@ -48,334 +67,334 @@ def get_indicators(self) -> dict:
...
+
+class DatasetBouquet(Base):
+    __tablename__ = "datasets_bouquets"
+
+    id: Mapped[int] = mapped_column(Integer, primary_key=True)
+    bouquet_id: Mapped[str] = mapped_column(String, ForeignKey("bouquets.bouquet_id"))
+    dataset_id: Mapped[str] = mapped_column(String, ForeignKey("catalog.dataset_id"))
+
+    def __repr__(self):
+        return f"<DatasetBouquet of <Bouquet {self.bouquet_id}> and <Dataset {self.dataset_id}>"
```
Comment (streino) [inline models.py:387]:
```suggestion
        return f"<DatasetBouquet of <Bouquet {self.bouquet_id}> and <Dataset {self.dataset_id}>>"
```

File: models.py [hunk 2]
```diff
@@ -48,334 +67,334 @@ def get_indicators(self) -> dict:
...
-    def from_record(cls, record: OrderedDict) -> DatasetRow:
-        return DatasetRow(**{k: v for k, v in record.items() if k != "id"})
 
-    @classmethod
-    def col_types(cls):
+class ResourceComputedColumns:
+    def __init__(self, payload: dict):
+        self.payload = payload
+
+    def get_indicators(self) -> dict:
```
Comment (streino) [inline models.py:244]: I'd be tempted to turn this into a static method of `Resource`. Same for `DatasetComputedColumns`. It's all stateless stuff (but for `DatasetCC` it'll have more impact).

Or at least make `XXXComputedColumns` a subclass of `XXX`, so the two are more clearly tied?

File: init.sql [hunk 3]
```diff
@@ -1,7 +1,7 @@
--- create a database for every env
+-- create a database for every env (match production names for easy restore)
 
-CREATE DATABASE demo;
-GRANT ALL PRIVILEGES ON DATABASE demo TO postgres;
+CREATE DATABASE dashboard_backend;
```
[the remark this answers is not in the document]
Comment (abulte) [reply init.sql:3]: Yep, but in another PR. This will require some dokku-fu to rename the existing demo database/service.

**Comment by abulte** [comment]
> Do I need to run actual tests on this or you're confident enough to go with just the code review?

Good question! I'm not super confident.

File: models.py (hunk 2 above)
Comment (streino) [reply models.py:240]: Good points about keeping it separate. And no we don't need to handle this now.

**Comment by streino** [comment]
> > Do I need to run actual tests on this or you're confident enough to go with just the code review?
> 
> Good question! I'm not super confident.

I'll try to run some tests this week then

**Comment by streino** [comment]
> I'll try to run some tests this week then

Change of plan, I won't have time to test... So let's go ahead with this making sure we have a backup plan :crossed_fingers:

**Review by streino — APPROVED** [review]

**Pull request merged**